bpf/lib/encap.h

```diff
@@ -123,26 +115,32 @@ encap_and_redirect_lxc(struct __ctx_buff *ctx,
 		       __u32 seclabel, __u32 dstid,
 		       const struct trace_ctx *trace)
 {
-	struct tunnel_value *tunnel __maybe_unused;
+	struct remote_endpoint_info *ep __maybe_unused;
 
 	if (tunnel_endpoint)
 		return __encap_and_redirect_lxc(ctx, tunnel_endpoint,
 						encrypt_key, seclabel, dstid,
 						trace);
 
-	tunnel = map_lookup_elem(&TUNNEL_MAP, key);
-	if (!tunnel)
+	if (key->family == ENDPOINT_KEY_IPV4) {
+		ep = lookup_ip4_remote_endpoint(key->ip4, key->cluster_id);
+	} else {
+		ep = lookup_ip6_remote_endpoint(&key->ip6, key->cluster_id);
+	}
```

Comment:
Drive-by comment, there's no need for an extra ipcache lookup here, because all callers already performed it (that's how `tunnel_endpoint` gets normally populated). E.g,, https://github.com/pippolo84/cilium/blob/pr%2Fpippolo84%2Fmerge-tunnel-map-into-ipcache-exp/bpf/bpf_lxc.c#L920-L928

I haven't explicitly checked the other functions below, but AFAIR the same applies there as well.

Reply:
I had the same thought when first looking at it. One subtlety is that the initial ipcache lookup uses the precise IP, while here we use the (fixed-size) PodCIDR (as that's what the caller passed in the `tunnel_key`).

So right now this is a 1-to-1 replacement. Without considering that the IPcache is LPM and we would have already matched the PodCIDR entry on the first lookup.

Reply:
Ah right, this is using the masked IP as key. My comment was based on the fact that being the ipcache an LPM map, the node entry would already been retrieved during the previous lookup. And we'll also need to get rid of the fixed mask to benefit from the additional flexibility provided by the ipcache.

Reply:
> Drive-by comment, there's no need for an extra ipcache lookup here, because all callers already performed it (that's how `tunnel_endpoint` gets normally populated).

Tried that approach in https://github.com/cilium/cilium/pull/37146 and it seems to work correctly.